geo/src/algorithm/k_nearest_concave_hull.rs

```diff
@@ -0,0 +1,348 @@
+{
+    type Scalar = T;
+    fn k_nearest_concave_hull(&self, k: u32) -> Polygon<T> {
+        let points: Vec<Point<T>> = self.iter().flat_map(|poly| poly.points_iter().map(|x| x.clone())).collect();
+        concave_hull(points, k)
+    }
+}
+
+fn concave_hull<T>(mut points: Vec<Point<T>>, k: u32) -> Polygon<T>
+where T: GeoFloat + RTreeNum    
+{
+    points.dedup();
+
+    let set_length = points.len();
+    if set_length <= 3 {
+        return Polygon::new(LineString::from(points), vec![]);
+    }
+    if k >= set_length as u32 {
+        return fall_back_hull(points);
+    }
+
+    let k_adjusted = adjust_k(k);
+
+    let first_point = get_first_point(&points).clone();
+    let mut hull = vec![first_point.clone()];
+
+    let mut current_point = first_point;
+    let mut dataset = rstar::RTree::bulk_load(points.clone());
+
+    dataset.remove(&first_point);
+    let mut prev_angle = T::zero();
+    let mut curr_step = 2;
+    while (current_point != first_point || curr_step == 2) && dataset.size() > 0 {
+        if curr_step == 5 {
+            dataset.insert(first_point.clone());
+        }
+
+        let mut nearest_points = get_nearest_points(&dataset, &current_point, k_adjusted);
+        sort_by_angle(&mut nearest_points, &current_point, prev_angle);
+
+        let selected = nearest_points.iter().find(|x| !intersects(&hull, &[&current_point, x]));
+
+        if let Some(sel) = selected {
+            current_point = **sel;
+            hull.push(current_point.clone());
+            prev_angle = get_angle(&[&hull[hull.len() - 1], &hull[hull.len() - 2]]);
+            dataset.remove(&current_point);
+
+            curr_step += 1;
+        } else {
+            return concave_hull(points, get_next_k(k_adjusted));
+        }
+    }
+
+    let poly = Polygon::new(LineString::from(hull), vec![]);
+
+    if points.iter().any(|&p| !point_inside(&p, &poly)) {
+        return concave_hull(points, get_next_k(k_adjusted));
+    }
+
+    poly
+}
+
+fn fall_back_hull<T>(points: Vec<Point<T>>) -> Polygon<T>
+where T: GeoFloat
+{
+    let multipoint = MultiPoint::from(points);
+    multipoint.convex_hull()
+}
+
+fn get_next_k(curr_k: u32) -> u32 {
+    max(curr_k + 1, ((curr_k as f32) * K_MULTIPLIER) as u32)
+}
+
+fn adjust_k(k: u32) -> u32 {
+    max(k, 3)
+}
+
+fn get_first_point<T>(point_set: &Vec<Point<T>>) -> &Point<T>
+where T: GeoFloat
+{
+    let mut min_y = T::max_value();
+    let mut min_i = usize::MAX;
+
+    for (index, p) in point_set.iter().enumerate() {
+        if p.y() < min_y {
+            min_y = p.y();
+            min_i = index;
+        }
+    }
+
+    &point_set[min_i]
+}
+
+fn get_nearest_points<'a, T>(dataset: &'a rstar::RTree<Point<T>>, base_point: &Point<T>, candidate_no: u32) -> Vec<&'a Point<T>>
+where T: GeoFloat + RTreeNum
+{
+    dataset.nearest_neighbor_iter(base_point).take(candidate_no as usize).collect()
+}
+
+fn sort_by_angle<'a, T>(points: &'a mut Vec<&Point<T>>, curr_point: &Point<T>, prev_angle: T)
```

Comment:
Would it be possible to use a cross product here (signed area) instead of computing an angle?

Reply:
I've tried using dot-product method, but finally I've found even faster algorithm that works just fine.